lib/grant-types/abstract-grant-type.ts

```diff
@@ -0,0 +1,108 @@
+import { MS_IN_S } from '../constants';
+import { InvalidArgumentError, InvalidScopeError } from '../errors';
+import { Client, Model, User } from '../interfaces';
+import { Request } from '../request';
+import * as tokenUtil from '../utils/token-util';
+import * as is from '../validator/is';
+
+export class AbstractGrantType {
+  accessTokenLifetime: number;
+  model: Model;
+  refreshTokenLifetime: number;
+  alwaysIssueNewRefreshToken: boolean;
+
+  constructor(options: any = {}) {
+    if (!options.accessTokenLifetime) {
+      throw new InvalidArgumentError(
+        'Missing parameter: `accessTokenLifetime`',
+      );
+    }
+
+    if (!options.model) {
+      throw new InvalidArgumentError('Missing parameter: `model`');
+    }
+
+    this.accessTokenLifetime = options.accessTokenLifetime;
+    this.model = options.model;
+    this.refreshTokenLifetime = options.refreshTokenLifetime;
+    this.alwaysIssueNewRefreshToken = options.alwaysIssueNewRefreshToken;
+  }
+
+  /**
+   * Generate access token.
+   */
+
+  async generateAccessToken(client?: Client, user?: User, scope?: string) {
+    if (this.model.generateAccessToken) {
+      const token = await this.model.generateAccessToken(client, user, scope);
```

Comment:
can await be added to both `async function`  and normal `function`?
why not still using `promisify` to be compatible?

Reply:
or 
```js
const token = await Promise.resolve(this.model.generateAccessToken(client, user, scope));
```

Reply:
> 
> 
> can await be added to both `async function` and normal `function`?

I think it will work, but `promisify` should do the trick

Reply:
node8 and above support native promise which are standard for handling asynchronous flow

Reply:
no need to wrap in Promise.resolve, await will have the same effect

Reply:
you are right.